requests-for-comments/2022-06-01-session-analysis.md

```diff
@@ -0,0 +1,118 @@
+# Sessions in insights
+
+The goal of this document is to propose a specific approach to how we add session analysis to our insights product as defined in the [Product Context document](https://github.com/PostHog/product-internal/pull/307/files),
+
+For more details on the problem we're trying to solve and the customer needs, please reference that document.
+
+**Note:** The goal of this project is not to build a feature that allows users to find and explore specific sessions. Rather this feature is targeted at doing analysis on sessions in aggregate
+
+## Requirements
+
+We need to figure out how to allow users to do “session based analysis” inside our product. As defined in the document above, this means allowing users to do analysis on the following:
+
+1. Number of sessions where specific events occur over time
+2. Average session length over time
+3. Average number of sessions per user over time
+4. Session length distributions
+5. Count of a specific event per session over time
+6. Limit a funnel conversion to a single session
+7. Limit a path conversion to a single session
+
+## Possible Solutions
+
+At it’s highest level, there are 2 options here:
+
+1. Fitting session analysis into our existing insights features (mainly trends)
+2. Building out a new sessions feature tailor made for this analysis
+
+(tl;dr we should do option 1 because we can lean on lots of existing behavior of trends)
+
+The big trade-offs between these two options are:
+
+- Option 1 allows for lots of analysis that you “get for free” because of the power of trends. For example the following are all easily possible:
+    - Plot the number of sessions over time along with the count of users. Then use formulas to divide the 2 and get “sessions per user”
+    - Plot average session length where a users does Z broken down by browser, and notice that your feature is broken on IE.
+    - Plot the average session length where users do X and compare it to the average session length where users do Y. Notice that users that do X, stick around longer.
+    - Run an experiment where the goal is to increase the session length 🤯
+- Potentially, option 2 would allow for more complex sessions to be defined. Right now, the language of trends is about defining events, not sessions. As a result, when this language is translated to sessions, it means, “Give me sessions where this specific event occurs”. This restricts users from creating complex session definitions such as “sessions where X happens and the Y happens within 2 minutes”.
```

Comment:
True, but the example is a funnel aggregated by sessions, so should be possible in the new view. We need another example of a complex session that isn't possible using any existing insights lol.

Reply:
valid example!

Aside: I'm not sure it requires something dedicated to solve this. For example, this sounds like a trend graph, with property aggregation of average, on the property session length, and filtered by a cohort of sessions created using the funnel.

Basically, the missing primitive is 'cohorts of groups', which will solve this problem for all groups, not just sessions.

But yeah, agree that these kinds of complex queries support aren't required yet.

![its-all-connected-everything-is-connected](https://user-images.githubusercontent.com/7115141/171656607-3100b144-e658-4903-8897-62a1e7ae91e0.gif)